File: Algorithms/78.子集/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        return list(self.subsets_from(nums, 0))
    
    def subsets_from(self, nums, k):
        if k == len(nums):
            yield []
        else:
            yield []
            for i in range(k, len(nums)):
                for subset in self.subsets_from(nums, i + 1):
                    yield [nums[i]] + subset

File: 程序员面试金典/8.4/subsets.java
// 方法1：迭代
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        ans.add(new ArrayList<>());
        for (int n : nums) {
            int size = ans.size();
            for (int i = 0; i < size; ++i) {
                List<Integer> temp = (ArrayList<Integer>) ((ArrayList<Integer>) ans.get(i)).clone();
                temp.add(n);
                ans.add(temp);
            }
        }
        return ans;
    }
}
